src/pyflakes/checks.rs

```diff
@@ -23,6 +24,210 @@ fn has_star_args(args: &[Expr]) -> bool {
+            | ExprKind::Tuple { .. }
+            | ExprKind::Set { .. }
+            | ExprKind::ListComp { .. }
+            | ExprKind::SetComp { .. }
+            | ExprKind::GeneratorExp { .. } => Some(Check::new(
+                CheckKind::PercentFormatExpectedMapping,
+                location,
+            )),
+            _ => None,
+        }
+    }
+}
+
+/// F503
+pub(crate) fn percent_format_expected_sequence(
+    summary: &CFormatSummary,
+    right: &Expr,
+    location: Range,
+) -> Option<Check> {
+    if summary.num_positional <= 1 {
+        None
+    } else {
+        match right.node {
+            ExprKind::Dict { .. } | ExprKind::DictComp { .. } => Some(Check::new(
+                CheckKind::PercentFormatExpectedSequence,
+                location,
+            )),
+            _ => None,
+        }
+    }
+}
+
+/// F504
+pub(crate) fn percent_format_extra_named_arguments(
+    summary: &CFormatSummary,
+    right: &Expr,
+    location: Range,
+) -> Option<Check> {
+    if summary.num_positional > 0 {
+        return None;
+    }
+
+    if let ExprKind::Dict { keys, .. } = &right.node {
+        let missing: Vec<String> = keys
+            .iter()
+            .filter_map(|k| match &k.node {
+                // We can only check that string literals exist
+                ExprKind::Constant {
+                    value: Constant::Str(value),
+                    ..
+                } => {
+                    if summary.keywords.contains(value) {
+                        None
+                    } else {
+                        Some(value)
+                    }
+                }
+                _ => None,
+            })
+            .cloned()
+            .collect();
+
+        if missing.is_empty() {
+            None
+        } else {
+            Some(Check::new(
+                CheckKind::PercentFormatExtraNamedArguments(missing),
+                location,
+            ))
+        }
+    } else {
+        None
+    }
+}
+
+/// F505
+pub(crate) fn percent_format_missing_arguments(
+    summary: &CFormatSummary,
+    right: &Expr,
+    location: Range,
+) -> Option<Check> {
+    if summary.num_positional > 0 {
+        return None;
+    }
+
+    if let ExprKind::Dict { keys, values } = &right.node {
+        if values.len() > keys.len() {
+            return None; // contains **x splat
+        }
+
+        let mut keywords = FxHashSet::default();
+        for key in keys {
+            match &key.node {
+                ExprKind::Constant {
+                    value: Constant::Str(value),
+                    ..
+                } => {
+                    keywords.insert(value);
+                }
+                _ => {
+                    return None; // Dynamic keys present
+                }
+            }
+        }
+
+        let missing: Vec<String> = summary
+            .keywords
+            .iter()
+            .filter(|k| !keywords.contains(k))
+            .cloned()
```

Comment:
We might be able to delay these `cloned` calls until the `else` block? Like make this `Vec<&str>` and then call `cloned` when creating (e.g.) `CheckKind::PercentFormatMissingArgument`? Really minor though (and didn't try it myself).

Reply:
Yeah, that's possible - we actually get a `Vec<&String>` and then put `missing.iter().map(|&s| s.clone()).collect()` into the check. 

I had assumed (possibly naively) that `cloned` would have zero overhead if the filter produces an empty iterator (which is what the if/else block is checking)?

Reply:
Unless the move is more about signalling intent than performance?

Reply:
Yeah I'm sure the difference is negligible if existent at all given that we're only cloning the empty `vec` anyway in that case. It's more about signaling intent, locally and broadly ("clone as late as possible"), but honestly don't feel strongly about changing it.

Reply:
Makes sense - I've moved cloning later in the 3 functions where it was clear how to do so